cpp/src/arrow/compute/kernel.h

```diff
@@ -348,7 +348,24 @@ class ARROW_EXPORT OutputType {
   Resolver resolver_ = NULLPTR;
 };
 
-/// \brief Holds the input types and output type of the kernel.
+/// \brief Additional constraints to apply to the input types of a kernel when matching a
+/// specific kernel signature.
+class ARROW_EXPORT MatchConstraint {
+ public:
+  virtual ~MatchConstraint() = default;
+
+  /// \brief Return true if the input types satisfy the constraint.
+  virtual bool Validate(const std::vector<TypeHolder>& types) const = 0;
```

Comment:
`Validate()` -> `Matches()`?

Reply:
Hmm, I don't think it's natural to say a constraint "matches" the given list of types?

Reply:
Updated.